linker/Linker.Steps/MarkStep.cs

```diff
@@ -2231,6 +2504,21 @@ public AttributeProviderPair (CustomAttribute attribute, ICustomAttributeProvide
 			public CustomAttribute Attribute { get; private set; }
 			public ICustomAttributeProvider Provider { get; private set; }
 		}
+		
+		protected class ActivatorCreateInstanceMarkingInformation {
+			public ActivatorCreateInstanceMarkingInformation (MethodBody callingBody, TypeDefinition castType, Func<TypeDefinition, bool, MethodDefinition[]> constructorCollector, bool defaultCtorOnly)
+			{
+				CallingBody = callingBody;
+				CastType = castType;
+				ConstructorCollector = constructorCollector;
+				DefaultCtorOnly = defaultCtorOnly;
+			}
+
+			public MethodBody CallingBody { get; private set; }
+			public TypeDefinition CastType { get; private set; }
+			public Func<TypeDefinition, bool, MethodDefinition[]> ConstructorCollector { get; private set; }
```

Comment:
Is it necessary to have this customizable for each instance?

Reply:
Yes.  Each `Activator.CreateInstance` could trigger 1 of a number of marking scenarios.

1) We know only the default ctor needs to be marked.

2) We you are create some instance of a type, but we have no idea which ctor you are trying to use, so mark them all.

3) We know you want a default ctor, but only if it's public or non-public (I didn't implement this distinction because it didn't seem worth the effort, but if a need ever arose this plumbing would be ready).  This scenario is encountered by the `CreateInstance(Type type, bool nonPublic)` overload

4) When you combine UnityLinker's stack evaluation with the more complex overloads of `CreateInstance` that let you specify `BindingFlags` and pass the ctor parameters, it's conceivable that if we had a very targeted case that we wanted to detect we could go so far as to match the types in the ctor params object[] array with the parameters of the ctor's on the type and only mark the ctor(s) that full fill the same signature.

1 & 2 are implemented in this PR.  3 & 4 may go unimplemented until a need arises, but I'd like to have the paradigm that is in place be ready to support these scenarios so that I (or someone else) doesn't have to revisit this

Reply:
Please don't make the code over engineered when the need comes we can refactor it but custom ConstructorCollector for each instance seems really unnecessary for this change

Reply:
My goal isn't to over engineer, it's to leave a door open.  Each time I need to make a change upstream it leads to a series of hoops I have to jump through on our end before I can complete w/e it is I set out to do.  If I might want access to something in the future, it's simply easier to leave myself an open door.

I haven't looked at how this code plays out on a real example yet and it's probably a good idea to give it a try. I linked a simple Unity project with these changes and here is what `_activatorCreateInstanceTypes` looked like each time `ProcessActivatorCreateInstanceType` was called.

```
Call 1 - _activatorCreateInstanceTypes.Count = 3
Call 2 - _activatorCreateInstanceTypes.Count = 3
Call 3 - _activatorCreateInstanceTypes.Count = 5
Call 4 - _activatorCreateInstanceTypes.Count = 6
Call 5 - _activatorCreateInstanceTypes.Count = 7
Call 6 - _activatorCreateInstanceTypes.Count = 8
Call 7 - _activatorCreateInstanceTypes.Count = 8
Call 8 - _activatorCreateInstanceTypes.Count = 8
Call 9 - _activatorCreateInstanceTypes.Count = 8
Call 10 - _activatorCreateInstanceTypes.Count = 8
```

I don't see the worry with the 10 `ConstructorCollector` Func's that are created. 

I would like to leave a door open so that I don't have to redo this dance again in the future.  An alternative I see that keeps the door open is the Func property changes to a bool property named `DefaultCtorOnly`, but 1 or 2 new methods are added for us to hook into the creation of and processing of an `ActivatorCreateInstanceMarkingInformation`

Would you prefer that approach?